File: src/cerebro_sdk/agents/tooling.py
"""Agent tooling and approval helpers."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Optional
from uuid import UUID

from prometheus_client import CollectorRegistry
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from cerebro.agents.models import AgentPolicySuggestion, ToolApproval, ToolInvocation

from cerebro_sdk.agents.base import AsyncManagerBase
from cerebro_sdk.agents.repositories import ToolingRepository
from cerebro_sdk.agents.types import (
    AgentInvalidStatusError,
    AgentNotFoundError,
    AgentPolicySuggestionRecord,
    ToolApprovalRecord,
    ToolInvocationRecord,
)

from cerebro.agents.models import ApprovalStatus, ToolInvocationStatus


class AgentToolingManager(AsyncManagerBase):
    """Inspect and manage tool invocations and approvals."""

    def __init__(self, db: AsyncSession, *, registry: CollectorRegistry | None = None) -> None:
        super().__init__(db)
        self._repo = ToolingRepository(db, registry=registry)

    async def list_invocations(
        self,
        *,
        session_id: Optional[UUID] = None,
        org_id: Optional[UUID] = None,
        status: ToolInvocationStatus | str | None = None,
        tool_name: Optional[str] = None,
        since: Optional[datetime] = None,
        until: Optional[datetime] = None,
        cursor: Optional[datetime] = None,
        page_size: Optional[int] = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[ToolInvocationRecord]:
        status_enum: Optional[ToolInvocationStatus] = None
        if status:
            try:
                status_enum = self._require_enum(
                    status,
                    ToolInvocationStatus,
                    message=f"Invalid tool invocation status '{status}'",
                )
            except AgentInvalidStatusError:
                return []

        effective_limit = page_size if page_size is not None else limit
        invocations = await self._repo.list_invocations(
            session_id=session_id,
            org_id=org_id,
            status=status_enum,
            tool_name=tool_name,
            since=since,
            until=until,
            cursor=cursor,
            effective_limit=effective_limit,
            offset=offset,
        )
        return [self._invocation_to_record(invocation) for invocation in invocations]

    async def get_invocation(self, invocation_id: UUID) -> Optional[ToolInvocationRecord]:
        invocation = await self._repo.get_invocation(invocation_id)
        if not invocation:
            return None
        return self._invocation_to_record(invocation)

    async def create_invocation(
        self,
        *,
        session_id: UUID,
        tool_name: str,
        tool_version: str = "1.0",
        input_data: dict[str, Any],
        status: ToolInvocationStatus | str = ToolInvocationStatus.PENDING,
        started_at: Optional[datetime] = None,
        cel_policy_key: Optional[str] = None,
        cel_expression: Optional[str] = None,
        cel_context: Optional[dict[str, Any]] = None,
    ) -> ToolInvocationRecord:
        status_enum = self._require_enum(
            status,
            ToolInvocationStatus,
            message=f"Invalid tool invocation status '{status}'",
        )
        started_at_value = started_at or datetime.now(timezone.utc)
        async with self._transaction():
            invocation = await self._repo.create_invocation(
                session_id=session_id,
                tool_name=tool_name,
                tool_version=tool_version,
                input_data=input_data,
                status=status_enum,
                started_at=started_at_value,
                cel_policy_key=cel_policy_key,
                cel_expression=cel_expression,
                cel_context=cel_context,
            )
        await self._db.refresh(invocation)
        return self._invocation_to_record(invocation)

    async def list_approvals(
        self,
        *,
        org_id: UUID,
        status: ApprovalStatus | str | None = None,
        since: Optional[datetime] = None,
        until: Optional[datetime] = None,
        cursor: Optional[datetime] = None,
        page_size: Optional[int] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[ToolApprovalRecord]:
        status_enum: Optional[ApprovalStatus] = None
        if status:
            try:
                status_enum = self._require_enum(
                    status,
                    ApprovalStatus,
                    message=f"Invalid approval status '{status}'",
                )
            except AgentInvalidStatusError:
                return []

        effective_limit = page_size if page_size is not None else limit
        approvals = await self._repo.list_approvals(
            org_id=org_id,
            status=status_enum,
            since=since,
            until=until,
            cursor=cursor,
            effective_limit=effective_limit,
            offset=offset,
        )
        return [self._approval_to_record(approval) for approval in approvals]

    async def update_approval_status(
        self,
        *,
        approval_id: UUID,
        status: ApprovalStatus | str,
        decided_by: str,
        decision_reason: Optional[str] = None,
    ) -> Optional[ToolApprovalRecord]:
        approval = await self._repo.get_approval(approval_id)
        if not approval:
            return None
        status_enum = self._require_enum(
            status,
            ApprovalStatus,
            message=f"Invalid approval status '{status}'",
        )

        invocation = approval.tool_invocation
        now = datetime.now(timezone.utc)

        async with self._transaction():
            approval.status = status_enum
            approval.decided_by = decided_by
            approval.decided_at = now
            approval.decision_reason = decision_reason

            if invocation:
                if status_enum == ApprovalStatus.APPROVED:
                    invocation.status = ToolInvocationStatus.SUCCESS
                    invocation.completed_at = now
                elif status_enum == ApprovalStatus.REJECTED:
                    invocation.status = ToolInvocationStatus.ERROR
                    invocation.error_message = decision_reason or "Rejected by reviewer"
                    invocation.completed_at = now

        await self._db.refresh(approval)
        if invocation:
            await self._db.refresh(invocation)
        return self._approval_to_record(approval)

    async def update_invocation_result(
        self,
        *,
        invocation_id: UUID,
        status: ToolInvocationStatus | str | None = None,
        output_data: Optional[dict[str, Any]] = None,
        error_message: Optional[str] = None,
        error_code: Optional[str] = None,
        cel_result: Optional[bool] = None,
        cel_context: Optional[dict[str, Any]] = None,
        completed_at: Optional[datetime] = None,
    ) -> ToolInvocationRecord:
        invocation = await self._repo.get_invocation(invocation_id)
        if not invocation:
            raise AgentNotFoundError(f"Tool invocation {invocation_id} not found")

        status_enum: Optional[ToolInvocationStatus] = None
        if status is not None:
            status_enum = self._require_enum(
                status,
                ToolInvocationStatus,
                message=f"Invalid tool invocation status '{status}'",
            )

        completed_value: Optional[datetime] = completed_at
        if completed_value is None and status_enum in {
            ToolInvocationStatus.SUCCESS,
            ToolInvocationStatus.ERROR,
            ToolInvocationStatus.DRY_RUN,
        }:
            completed_value = datetime.now(timezone.utc)

        async with self._transaction():
            await self._repo.update_invocation(
                invocation,
                status=status_enum,
                output_data=output_data,
                error_message=error_message,
                error_code=error_code,
                completed_at=completed_value,
                cel_result=cel_result,
                cel_context=cel_context,
            )

        await self._db.refresh(invocation)
        return self._invocation_to_record(invocation)

    async def list_policy_suggestions(
        self,
        *,
        org_id: UUID,
        tool_name: Optional[str] = None,
    ) -> list[AgentPolicySuggestionRecord]:
        stmt = select(AgentPolicySuggestion).where(AgentPolicySuggestion.org_id == org_id)
        if tool_name:
            stmt = stmt.where(AgentPolicySuggestion.tool_name == tool_name)
        stmt = stmt.order_by(AgentPolicySuggestion.last_seen.desc())
        suggestions = list(await self._db.scalars(stmt))
        return [self._policy_to_record(suggestion) for suggestion in suggestions]

    @staticmethod
    def _invocation_to_record(invocation: ToolInvocation) -> ToolInvocationRecord:
        input_payload = invocation.input_data if isinstance(invocation.input_data, dict) else {"value": invocation.input_data}
        output_payload = None
        if invocation.output_data is not None:
            output_payload = invocation.output_data if isinstance(invocation.output_data, dict) else {"value": invocation.output_data}
        return ToolInvocationRecord(
            invocation_id=invocation.id,
            session_id=invocation.session_id,
            tool_name=invocation.tool_name,
            tool_version=invocation.tool_version,
            status=invocation.status.value,
            started_at=invocation.started_at,
            completed_at=invocation.completed_at,
            input_data=input_payload,
            output_data=output_payload,
            error_message=invocation.error_message,
            cel_policy_key=invocation.cel_policy_key,
            cel_expression=invocation.cel_expression,
            cel_result=invocation.cel_result,
        )

    @staticmethod
    def _approval_to_record(approval: ToolApproval) -> ToolApprovalRecord:
        return ToolApprovalRecord(
            approval_id=approval.id,
            org_id=approval.org_id,
            tool_invocation_id=approval.tool_invocation_id,
            requested_by=approval.requested_by,
            requested_at=approval.requested_at,
            reason=approval.reason,
            status=approval.status.value,
            decided_by=approval.decided_by,
            decided_at=approval.decided_at,
            decision_reason=approval.decision_reason,
            expires_at=approval.expires_at,
            risk_assessment=approval.risk_assessment if isinstance(approval.risk_assessment, dict) else {},
        )

    @staticmethod
    def _policy_to_record(suggestion: AgentPolicySuggestion) -> AgentPolicySuggestionRecord:
        return AgentPolicySuggestionRecord(
            suggestion_id=suggestion.id,
            org_id=suggestion.org_id,
            tool_name=suggestion.tool_name,
            cel_expression=suggestion.cel_expression,
            support_count=suggestion.support_count,
            reject_count=suggestion.reject_count,
            details=suggestion.details if isinstance(suggestion.details, dict) else {},
            last_seen=suggestion.last_seen,
            created_at=suggestion.created_at,
        )

File: sdk/ts/src/clients/tooling.ts
import HttpClient from "../httpClient.js";
import {
  PolicySuggestionRecord,
  ToolApprovalRecord,
  ToolInvocationRecord,
  ToolInvocationSummary,
} from "../types.js";
import { AgentsClient } from "./agents.js";

export interface ListToolInvocationsOptions {
  status?: string;
  toolName?: string;
  limit?: number;
  since?: string | Date;
}

export interface CreateToolInvocationRequest {
  sessionId: string;
  toolName: string;
  toolVersion?: string;
  inputData: Record<string, unknown>;
  celPolicyKey?: string;
}

export interface UpdateToolInvocationOptions {
  invocationId: string;
  status?: string;
  outputData?: Record<string, unknown> | null;
  errorMessage?: string | null;
  errorCode?: string | null;
  celResult?: boolean | null;
  celContext?: Record<string, unknown> | null;
}

export interface ListToolApprovalsOptions {
  status?: string;
  limit?: number;
}

export interface UpdateApprovalStatusRequest {
  approvalId: string;
  status: string;
  decidedBy: string;
  decisionReason?: string | null;
}

export interface ToolingAdapter {
  listInvocations(options?: ListToolInvocationsOptions): Promise<ToolInvocationRecord[]>;
  getInvocation?(invocationId: string): Promise<ToolInvocationRecord | null>;
  createInvocation?(request: CreateToolInvocationRequest): Promise<ToolInvocationRecord>;
  updateInvocationResult?(options: UpdateToolInvocationOptions): Promise<ToolInvocationRecord>;
  listApprovals?(options?: ListToolApprovalsOptions): Promise<ToolApprovalRecord[]>;
  updateApprovalStatus?(request: UpdateApprovalStatusRequest): Promise<ToolApprovalRecord>;
  summarizeInvocations?(options?: ListToolInvocationsOptions): Promise<ToolInvocationSummary[]>;
}

export interface HttpToolingAdapterOptions {
  basePath?: string;
}

interface ToolInvocationPayload {
  id: string;
  session_id?: string;
  tool_name: string;
  tool_version?: string;
  status: string;
  started_at: string;
  completed_at: string | null;
  error_message?: string | null;
  error_code?: string | null;
  input_data?: Record<string, unknown> | null;
  output_data?: Record<string, unknown> | null;
  cel_policy_key?: string | null;
  cel_expression?: string | null;
  cel_result?: boolean | null;
  cel_context?: Record<string, unknown> | null;
}

interface ToolApprovalPayload {
  id: string;
  org_id: string;
  tool_invocation_id: string;
  requested_by: string;
  requested_at: string;
  reason: string;
  status: string;
  decided_by?: string | null;
  decided_at?: string | null;
  decision_reason?: string | null;
  expires_at?: string | null;
  risk_assessment?: Record<string, unknown> | null;
}

interface ToolInvocationSummaryPayload {
  tool_name: string;
  status: string;
  count: number;
}

class HttpToolingAdapter implements ToolingAdapter {
  constructor(
    private readonly http: HttpClient,
    private readonly options: HttpToolingAdapterOptions = {},
  ) {}

  private get basePath(): string {
    return this.options.basePath ?? "/api/v1/agents/tooling";
  }

  async listInvocations(options: ListToolInvocationsOptions = {}): Promise<ToolInvocationRecord[]> {
    const params: Record<string, string | number> = {};
    if (options.status) params.status = options.status;
    if (options.toolName) params.tool_name = options.toolName;
    if (options.limit !== undefined) params.limit = options.limit;
    if (options.since) params.since = normaliseDate(options.since);

    const payload = await this.http.get<ToolInvocationPayload[]>(`${this.basePath}/invocations`, {
      searchParams: Object.keys(params).length ? params : undefined,
    });

    return payload.map(mapInvocation);
  }

  async getInvocation(invocationId: string): Promise<ToolInvocationRecord | null> {
    const payload = await this.http.get<ToolInvocationPayload | null>(`${this.basePath}/invocations/${invocationId}`);
    return payload ? mapInvocation(payload) : null;
  }

  async createInvocation(request: CreateToolInvocationRequest): Promise<ToolInvocationRecord> {
    const payload = await this.http.post<ToolInvocationPayload>(`${this.basePath}/invocations`, {
      body: {
        session_id: request.sessionId,
        tool_name: request.toolName,
        tool_version: request.toolVersion,
        input_data: request.inputData,
        cel_policy_key: request.celPolicyKey,
      },
    });

    return mapInvocation(payload);
  }

  async updateInvocationResult(options: UpdateToolInvocationOptions): Promise<ToolInvocationRecord> {
    const payload = await this.http.post<ToolInvocationPayload>(`${this.basePath}/invocations/${options.invocationId}/result`, {
      body: {
        status: options.status,
        output_data: options.outputData,
        error_message: options.errorMessage,
        error_code: options.errorCode,
        cel_result: options.celResult,
        cel_context: options.celContext,
      },
    });

    return mapInvocation(payload);
  }

  async listApprovals(options: ListToolApprovalsOptions = {}): Promise<ToolApprovalRecord[]> {
    const params: Record<string, string | number> = {};
    if (options.status) params.status = options.status;
    if (options.limit !== undefined) params.limit = options.limit;

    const payload = await this.http.get<ToolApprovalPayload[]>(`${this.basePath}/approvals`, {
      searchParams: Object.keys(params).length ? params : undefined,
    });

    return payload.map(mapApproval);
  }

  async updateApprovalStatus(request: UpdateApprovalStatusRequest): Promise<ToolApprovalRecord> {
    const payload = await this.http.post<ToolApprovalPayload>(`${this.basePath}/approvals/${request.approvalId}`, {
      body: {
        status: request.status,
        decided_by: request.decidedBy,
        decision_reason: request.decisionReason ?? null,
      },
    });

    return mapApproval(payload);
  }

  async summarizeInvocations(options: ListToolInvocationsOptions = {}): Promise<ToolInvocationSummary[]> {
    const params: Record<string, string | number> = {};
    if (options.status) params.status = options.status;
    if (options.toolName) params.tool_name = options.toolName;

    const payload = await this.http.get<ToolInvocationSummaryPayload[]>(`${this.basePath}/invocations/summary`, {
      searchParams: Object.keys(params).length ? params : undefined,
    });

    return payload.map((entry) => ({
      toolName: entry.tool_name,
      status: entry.status,
      count: entry.count,
    }));
  }
}

export class InMemoryToolingAdapter implements ToolingAdapter {
  private readonly invocations = new Map<string, ToolInvocationRecord>();
  private readonly approvals = new Map<string, ToolApprovalRecord>();

  async listInvocations(): Promise<ToolInvocationRecord[]> {
    return Array.from(this.invocations.values());
  }

  async getInvocation(invocationId: string): Promise<ToolInvocationRecord | null> {
    return this.invocations.get(invocationId) ?? null;
  }

  async createInvocation(request: CreateToolInvocationRequest): Promise<ToolInvocationRecord> {
    const record: ToolInvocationRecord = {
      invocationId: generateId(),
      sessionId: request.sessionId,
      toolName: request.toolName,
      toolVersion: request.toolVersion,
      status: "PENDING",
      startedAt: new Date(),
      completedAt: null,
      errorMessage: null,
      inputData: request.inputData,
      outputData: null,
      errorCode: null,
      celPolicyKey: request.celPolicyKey ?? null,
      celExpression: null,
      celResult: null,
      celContext: null,
    };
    this.invocations.set(record.invocationId, record);
    return record;
  }

  async updateInvocationResult(options: UpdateToolInvocationOptions): Promise<ToolInvocationRecord> {
    const existing = this.invocations.get(options.invocationId);
    if (!existing) {
      throw new Error(`Invocation '${options.invocationId}' not found`);
    }
    if (options.status) existing.status = options.status;
    if (options.outputData !== undefined) existing.outputData = options.outputData;
    if (options.errorMessage !== undefined) existing.errorMessage = options.errorMessage;
    if (options.errorCode !== undefined) existing.errorCode = options.errorCode;
    if (options.celResult !== undefined) existing.celResult = options.celResult;
    if (options.celContext !== undefined) existing.celContext = options.celContext;
    existing.completedAt = new Date();
    return existing;
  }

  async listApprovals(): Promise<ToolApprovalRecord[]> {
    return Array.from(this.approvals.values());
  }

  async updateApprovalStatus(request: UpdateApprovalStatusRequest): Promise<ToolApprovalRecord> {
    const approval = this.approvals.get(request.approvalId);
    if (!approval) {
      throw new Error(`Approval '${request.approvalId}' not found`);
    }
    approval.status = request.status;
    approval.decidedBy = request.decidedBy;
    approval.decidedAt = new Date();
    approval.decisionReason = request.decisionReason ?? null;
    return approval;
  }

  async summarizeInvocations(): Promise<ToolInvocationSummary[]> {
    const summary = new Map<string, number>();
    for (const invocation of this.invocations.values()) {
      const key = `${invocation.toolName}:${invocation.status}`;
      summary.set(key, (summary.get(key) ?? 0) + 1);
    }

    return Array.from(summary.entries()).map(([key, count]) => {
      const [toolName, status] = key.split(":");
      return { toolName, status, count };
    });
  }
}

export class AgentToolingClient {
  constructor(
    private readonly agents: AgentsClient,
    private readonly adapter?: ToolingAdapter,
  ) {}

  static fromHttpClient(http: HttpClient, agents: AgentsClient, options?: HttpToolingAdapterOptions): AgentToolingClient {
    return new AgentToolingClient(agents, new HttpToolingAdapter(http, options));
  }

  async listPolicySuggestions(options?: { limit?: number }): Promise<PolicySuggestionRecord[]> {
    return this.agents.listPolicySuggestions(options);
  }

  async simulatePolicyExpression(request: { expression: string; toolName?: string; limit?: number }) {
    return this.agents.simulatePolicyExpression(request);
  }

  async listInvocations(options?: ListToolInvocationsOptions): Promise<ToolInvocationRecord[]> {
    this.ensureAdapter("listInvocations");
    return this.adapter!.listInvocations(options);
  }

  async getInvocation(invocationId: string): Promise<ToolInvocationRecord | null> {
    this.ensureAdapter("getInvocation");
    return this.adapter!.getInvocation!(invocationId);
  }

  async createInvocation(request: CreateToolInvocationRequest): Promise<ToolInvocationRecord> {
    this.ensureAdapter("createInvocation");
    return this.adapter!.createInvocation!(request);
  }

  async updateInvocationResult(options: UpdateToolInvocationOptions): Promise<ToolInvocationRecord> {
    this.ensureAdapter("updateInvocationResult");
    return this.adapter!.updateInvocationResult!(options);
  }

  async listApprovals(options?: ListToolApprovalsOptions): Promise<ToolApprovalRecord[]> {
    this.ensureAdapter("listApprovals");
    return this.adapter!.listApprovals!(options);
  }

  async updateApprovalStatus(request: UpdateApprovalStatusRequest): Promise<ToolApprovalRecord> {
    this.ensureAdapter("updateApprovalStatus");
    return this.adapter!.updateApprovalStatus!(request);
  }

  async summarizeInvocations(options?: ListToolInvocationsOptions): Promise<ToolInvocationSummary[]> {
    this.ensureAdapter("summarizeInvocations");
    return this.adapter!.summarizeInvocations!(options);
  }

  private ensureAdapter(method: keyof ToolingAdapter): void {
    if (!this.adapter || !(method in this.adapter) || typeof (this.adapter as any)[method] !== "function") {
      throw new Error(`Tooling adapter does not implement '${method}'`);
    }
  }
}

function mapInvocation(payload: ToolInvocationPayload): ToolInvocationRecord {
  return {
    invocationId: payload.id,
    sessionId: payload.session_id,
    toolName: payload.tool_name,
    toolVersion: payload.tool_version,
    status: payload.status,
    startedAt: new Date(payload.started_at),
    completedAt: payload.completed_at ? new Date(payload.completed_at) : null,
    errorMessage: payload.error_message ?? null,
    errorCode: payload.error_code ?? null,
    inputData: payload.input_data ?? undefined,
    outputData: payload.output_data ?? undefined,
    celPolicyKey: payload.cel_policy_key ?? null,
    celExpression: payload.cel_expression ?? null,
    celResult: payload.cel_result ?? null,
    celContext: payload.cel_context ?? null,
  };
}

function mapApproval(payload: ToolApprovalPayload): ToolApprovalRecord {
  return {
    approvalId: payload.id,
    orgId: payload.org_id,
    toolInvocationId: payload.tool_invocation_id,
    requestedBy: payload.requested_by,
    requestedAt: new Date(payload.requested_at),
    reason: payload.reason,
    status: payload.status,
    decidedBy: payload.decided_by ?? null,
    decidedAt: payload.decided_at ? new Date(payload.decided_at) : null,
    decisionReason: payload.decision_reason ?? null,
    expiresAt: payload.expires_at ? new Date(payload.expires_at) : null,
    riskAssessment: payload.risk_assessment ?? {},
  };
}

function normaliseDate(value: string | Date): string {
  return value instanceof Date ? value.toISOString() : new Date(value).toISOString();
}

function generateId(): string {
  return Math.random().toString(36).slice(2, 10);
}
